text/0000-wire-reform.md

```diff
@@ -0,0 +1,146 @@
+* `@lwc/wire-service` becomes lightning platform specific for most part (`register()` method), since anyone can implement the wire adapter protocol.
+
+### Wire Adapter Protocol
+
+The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware for the component system, or even the application framework, it only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
+
+```ts
+interface WireAdapter {
+    update(config: Record<string, any>);
+    context(uid: string, value: any);
+    connect();
+    disconnect();
+}
+
+interface WireAdapterConstructor {
+    new (callback: emitDataCallback, contextualizer?: requestContextCallback): WireAdapter;
+}
+
+type emitDataCallback = (value: any) => void;
+
+type requestContextCallback = (uid: string) => void;
+```
+
+_Notes:_
+
+* Not all environments will support or need contextualization (e.g.: preloading LDS data), the `requestContextCallback` is optional. All adapters should be able to work with and without a context provider.
+
+## Interop
+
+If you have an adapter, you should be able to use it with any component system, not just LWC. This is an example of how to use this with React:
+
+```js
+// shared adapter
+import { MyWireAdapter } from 'some-module';
+
+class Foo extends React.Component {
+    constructor(props) {
+        super(props);
+        // The wire adapter instance is bound to the host object via the callback for data
+        this.adapter = new MyWireAdapter((data) => {
+            this.setState(() => {
+                // stream of data from wire adapter to be used to update the component's state
+                return data;
+            });
+        });
+        // calling for the initial update of the config since componentDidUpdate() is not
+        // invoked for the first time, but all props are ready.
+        this.adapter.update({ x: 1, y: this.props.valueOfY });
+    }
+
+    componentDidUpdate(prevProps) {
+        if (this.props.valueOfY !== prevProps.valueOfY) {
+            // recompute the config by extracting `this.props.valueOfY`
+            this.adapter.update({ x: 1, y: this.props.valueOfY });
+        }
+    }
+
+    componentDidMount() {
+        this.adapter.connect();
+    }
+
+    componentWillUnmount() {
+        this.adapter.disconnect();
+    }
+
+    render() {
+        return (
+            <div>{this.state.valueProducedByMyWireAdapter}</div>
+        );
+    }
+}
+```
+
+# Adoption strategy
+
+Since there is a need to support callable adapters that behave differently depending on who uses that (wire adapter vs user invoking the function directly), we have added a simple mechanism to support such feature via `adapter` property member expression on the callable. This opens the door to transition existing adapters to the new form. E.g.: APEX adapters are all callable objects.
```

Comment:
If I understand correctly, the Apex adapter implementation would look like:

```ts
export function invokeApex(...args) {
  if (new.target) {
    // invocation via new, return a WireAdapter instance
    const [dataCallback, contextCallback] = args;
  } else {
     // standard function call, return a Promise of some Apex controller result
    const [apexControllerParams] = args;
  }
}
```

Reply:
exactly! but in the interim, you can just continue doing what you're doing, by calling register on the wire-service package, and we can figure a way to do it for now for legacy adapters.